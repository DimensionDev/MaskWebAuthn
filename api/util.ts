import { decode } from 'cbor-redux'
import type { AttestationObject } from '../types/interface'
import { Buffer } from 'buffer'

export function concatenate(...arrays: ArrayBuffer[]): ArrayBuffer {
    const buffersLengths = arrays.map(function (b) {
        return b.byteLength
    })
    const totalLength = buffersLengths.reduce((p, c) => p + c, 0)
    const buffer = Buffer.alloc(totalLength)
    buffersLengths.reduce(function (p, c, i) {
        buffer.set(Buffer.from(arrays[i]), p)
        return p + c
    }, 0)
    return buffer.buffer
}

export async function verify(publicKey: CryptoKey, credential: PublicKeyCredential): Promise<boolean> {
    const { clientDataJSON, attestationObject } = credential.response as AuthenticatorAttestationResponse
    if (attestationObject) {
        const decodedAttestationObject = decode(attestationObject) as AttestationObject
        const {
            attStmt: { sig },
            antData,
        } = decodedAttestationObject
        const clientDataJsonHash = await crypto.subtle.digest('SHA-256', clientDataJSON)
        return crypto.subtle.verify(
            {
                name: 'ECDSA',
                hash: 'SHA-256',
            },
            publicKey,
            sig,
            concatenate(antData, clientDataJsonHash),
        )
    }
    // fallback
    return Promise.resolve().then(() => true)
}
