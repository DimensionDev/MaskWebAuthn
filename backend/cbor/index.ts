const keyToCOSEKey = (key: keyof JsonWebKey): [label: number, allowedType: Type[]] => {
    switch (key) {
        case 'kty':
            return [1, [MajorType.UTF8String, MajorType.PosInt, MajorType.NegInt]]
        case 'alg':
            return [3, [MajorType.ByteString]]
        case 'key_ops':
            return [4, [MajorType.UTF8String, MajorType.PosInt, MajorType.NegInt]]
        case 'crv':
            return [-1, [MajorType.UTF8String, MajorType.PosInt, MajorType.NegInt]]
        case 'x':
            return [-2, [MajorType.ByteString]]
        case 'y':
            return [-3, [MajorType.ByteString, SimpleType.True, SimpleType.False]]
        case 'd':
            return [-4, [MajorType.ByteString]]
        default:
            throw new TypeError()
    }
}

enum kty {
    OKP = 1,
    EC2 = 2,
    Symmetric = 4,
    Reserved = 0,
}

const INT8_MAX = 2 ** 8 - 1
const INT16_MAX = 2 ** 16 - 1
const INT32_MAX = 2 ** 32 - 1
const INF_MAP_START = 0xbf
const BREAK = 0xff

export enum MajorType {
    PosInt = 0b000, // 0
    NegInt = 0b001, // 1
    ByteString = 0b010, // 2
    UTF8String = 0b011, // 3
    Array = 0b100, // 4
    Map = 0b101, // 5
    Tag = 0b110, // 6
    Float = 0b111, // 7
}

export enum SimpleType {
    False = 0xf4, // 20
    True = 0xf5, // 21
    NULL = 0xf6, //22
    Undefined = 0xf7, //23
}

export type Type = MajorType | SimpleType

export const isMajorType = (v: Type): v is MajorType => v < 7
export const isSimpleType = (v: Type): v is SimpleType => v >= SimpleType.False && v <= SimpleType.Undefined

type WithType = (type: MajorType, payload: number) => number
type WithTypeWrapper = (payload: number) => ReturnType<WithType>

const withType: WithType = (type: MajorType, payload: number): number => (type << 5) + payload
const withNegInt: WithTypeWrapper = (payload: number) => withType(MajorType.NegInt, payload)
const withPosInt: WithTypeWrapper = (payload: number) => withType(MajorType.PosInt, payload)
const withUtf8: WithTypeWrapper = (payload: number) => withType(MajorType.UTF8String, payload)
const withBStr: WithTypeWrapper = (payload) => withType(MajorType.ByteString, payload)

export const toByteArray = (x: number, padding?: number): ArrayBuffer => {
    const LogTable = [1, 1, 2, 4, 4] as const
    const array = [] as Uint8Array[]
    if (x > INT32_MAX) {
        // javascript cannot handle bit operators with number larger than int32
        const y = Math.floor(x / 2 ** 32) // handle the high level
        array.unshift(Buffer.from(toByteArray(y)))
        while (array.length < 4) {
            array.unshift(Buffer.from([0]))
        }
        return Buffer.concat([...array, Buffer.from(toByteArray(x & INT32_MAX, 4))])
    } else {
        while (x > 0) {
            const byte = x & 0xff /* INT8_MAX */
            array.unshift(Buffer.from([byte]))
            x >>>= 8
        }
        while (array.length < (padding || LogTable[array.length])) {
            array.unshift(Buffer.from([0]))
        }
    }
    return Buffer.concat(array)
}

export const Number = (number: number): ArrayBuffer => {
    if (number >= 0 || number === -0) {
        return parseNumber(number, withPosInt)
    } else {
        return parseNumber(-number - 1, withNegInt)
    }
}

function parseNumber(number: number, withType: WithTypeWrapper): ArrayBuffer {
    let startBuffer: Buffer
    let endBuffer = Buffer.alloc(0)
    if (number < 0x18) {
        startBuffer = Buffer.from([withType(number)])
    } else {
        if (number <= INT8_MAX) {
            startBuffer = Buffer.from([withType(0x18)])
        } else if (number <= INT16_MAX) {
            startBuffer = Buffer.from([withType(0x19)])
        } else if (number <= INT32_MAX) {
            startBuffer = Buffer.from([withType(0x1a)])
        } else {
            startBuffer = Buffer.from([withType(0x1b)])
        }
        endBuffer = Buffer.from(toByteArray(number))
    }
    return Buffer.concat([startBuffer, endBuffer])
}

export const UTF8String = (string: string): ArrayBuffer => {
    const stringBuffer = Buffer.from(string, 'utf-8')
    const startBuffer = Buffer.from(parseNumber(stringBuffer.byteLength, withUtf8))
    return Buffer.concat([startBuffer, stringBuffer])
}

export const ByteString = (string: string): ArrayBuffer => {
    const stringBuffer = Buffer.from(string, 'hex')
    const startBuffer = Buffer.from(parseNumber(stringBuffer.byteLength, withBStr))
    return Buffer.concat([startBuffer, stringBuffer])
}

declare global {
    interface ObjectConstructor {
        getOwnPropertyNames<T extends any = any>(value: T): (keyof T)[]
    }
}

export const MajorTypeMap = {
    [MajorType.PosInt]: Number,
    [MajorType.NegInt]: Number,
    [MajorType.UTF8String]: UTF8String,
    [MajorType.ByteString]: ByteString,
} as Record<MajorType, <T extends any = any>(value: T) => ArrayBuffer>

export const SimpleTypeMap = {
    [SimpleType.True]: (value: any) => typeof value === 'boolean',
    [SimpleType.False]: (value: any) => typeof value === 'boolean',
    [SimpleType.Undefined]: (value: any) => value === undefined,
    [SimpleType.NULL]: (value: any) => value === null,
} as Record<SimpleType, <T extends any = any>(value: any) => value is T>

function parseType(types: Type[], value: any): ArrayBuffer {
    for (let type of types) {
        if (isMajorType(type)) {
            const parsed = matchMajorType(type, value)
            if (parsed != null) {
                return parsed
            }
        } else if (isSimpleType(type)) {
            const parsed = matchSimpleType(type, value)
            if (parsed != null) {
                return parsed
            }
        } else {
            throw new TypeError('unreachable')
        }
    }
    throw new TypeError('incorrect type')
    function matchMajorType(type: MajorType, value: any): undefined | ArrayBuffer {
        const callee = MajorTypeMap[type]
        if (callee) {
            return callee(value)
        } else {
            return undefined
        }
    }
    function matchSimpleType(type: SimpleType, value: boolean | undefined | null): undefined | ArrayBuffer {
        if (SimpleTypeMap[type](value)) {
            return Buffer.from([type])
        } else {
            return
        }
    }
}

// todo: refactor to parseObject
export function parseJsonWebKey(jwk: JsonWebKey): ArrayBuffer {
    const result: Buffer[] = []
    // todo: remove INF_MAP_START
    result.push(Buffer.from([INF_MAP_START]))
    for (let key of Object.getOwnPropertyNames(jwk)) {
        const [label, types] = keyToCOSEKey(key)
        pushKeyValue(Buffer.from(Number(label)), Buffer.from(parseType(types, jwk[key])))
    }

    // todo: remove INF_MAP_START
    result.push(Buffer.from([BREAK]))

    return Buffer.concat(result)
    function pushKeyValue(key: Buffer, value: Buffer) {
        result.push(key)
        result.push(value)
    }
}
