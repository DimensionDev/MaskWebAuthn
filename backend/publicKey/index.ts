export { create } from './create'
export { get } from './get'
