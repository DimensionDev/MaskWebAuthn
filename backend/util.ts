import type { CollectedClientData } from '../types/interface'
import { encode } from 'cbor-redux'

export function isRegistrableDomain(hostSuffixString: string, originalHost: string): boolean {
    // refs:
    //  https://html.spec.whatwg.org/multipage/origin.html#is-a-registrable-domain-suffix-of-or-is-equal-to
    //  https://github.com/passwordless-lib/fido2-net-lib/blob/bdad59ec9963c45c07b4c50b95cc3209d763a91e/Src/Fido2/AuthenticatorResponse.cs#L58-L83
    if (hostSuffixString === '' && originalHost === '') {
        return false
    }
    let host: string = ''
    let origin: string = ''
    try {
        const url = new URL(hostSuffixString)
        if (url.protocol !== 'https:') {
            return false
        }
        host = url.host
    } catch (_) {
        host = hostSuffixString
    }
    try {
        const url = new URL(originalHost)
        if (url.protocol !== 'https:') {
            return false
        }
        origin = url.host
    } catch (_) {
        origin = originalHost
    }

    if (origin.startsWith('localhost')) {
        // allow localhost
        return true
    } else {
        return host.endsWith(origin)
    }
}

export function checkUserVerification(userVerification: UserVerificationRequirement): boolean {
    switch (userVerification) {
        case 'discouraged':
            return false
        case 'preferred':
        case 'required':
        default:
            return true
    }
}

export function filterCredentials(credentials: PublicKeyCredentialDescriptor[]): PublicKeyCredentialDescriptor[] {
    return credentials.filter((credential) => {
        if (credential.transports && Array.isArray(credential.transports) && credential.transports.length > 0) {
            return false
        } else {
            return credential.type === 'public-key'
        }
    })
}

export function serializeCollectedClientData(collectedClientData: CollectedClientData): string {
    let result = ''
    result += '{'
    result += '"type":'
    result += ccdToString(collectedClientData.type)
    result += ',"challenge":'
    result += ccdToString(collectedClientData.challenge)
    result += ',"origin":'
    result += ccdToString(collectedClientData.origin)
    result += ',"crossOrigin":'
    result += collectedClientData.crossOrigin ? 'true' : 'false'
    // we don't handle the rest of the client data
    result += '}'
    return result
}

/**
 * @link https://www.w3.org/TR/webauthn-3/#ccdtostring
 */
export function ccdToString(obj: any) {
    let encoded = ''
    encoded += '"'
    const objString = `${obj}`
    for (const char of objString) {
        // check whether char is UTF-16 text
        // if `char.length > 1`, then it is the UTF-16
        const charCode: number =
            char.length > 1
                ? parseInt(char.charCodeAt(0).toString(16) + char.charCodeAt(1).toString(16), 16)
                : char.charCodeAt(0)
        // 0x20 space
        // 0x21 !
        // 0x22 "
        // 0x5c \
        if (
            charCode === 0x0020 ||
            charCode === 0x0021 ||
            (charCode >= 0x0023 && charCode <= 0x005b) ||
            (charCode >= 0x005d && charCode <= 0x10ffff)
        ) {
            encoded += char
        } else if (charCode === 0x22) {
            encoded += String.fromCharCode(0x5c, 0x22) // \"
        } else if (charCode === 0x5c) {
            encoded += String.fromCharCode(0x5c, 0x5c) // \\
        } else {
            encoded += '\\u' + charCode.toString(16) // \uxxxx
        }
    }
    encoded += '"'
    return encoded
}

export async function sha256(message: ArrayBuffer): Promise<ArrayBuffer> {
    return crypto.subtle.digest('SHA-256', message)
}

export enum AuthDataFlag {
    ED = 1 << 7,
    AT = 1 << 6,
    UV = 1 << 2,
    UP = 1 << 0,
}

export type AuthData = {
    rpIdHash: ArrayBuffer // sha256 hashed replying party id
    flags: AuthDataFlag
    signCount: number
    attestedCredentialData: {
        aaugid: string // is zero
        credentialId: ArrayBuffer
        credentialPublicKey: JsonWebKey
    }
    extensions: unknown // not support yet
}

export function stringToArrayBuffer(string: string, encoding: 'utf-8' | 'hex' = 'utf-8'): ArrayBuffer {
    if (encoding === 'utf-8') {
        const ret = new Uint8Array(string.length)
        for (let i = 0; i < string.length; i++) {
            ret[i] = string.charCodeAt(i)
        }
        return ret.buffer
    } else if (encoding === 'hex') {
        let length: number
        if (string.charAt(1) === 'x') {
            length = string.length - 2
        } else {
            length = string.length
        }
        const ret = new Uint8Array(length / 2)
        for (let i = 0; i < length / 2; i++) {
            const value = +string[i * 2] * 16 + +string[i * 2 + 1]
            ret.set([value], i * 2)
        }
        return ret
    } else {
        throw new Error('UNREACHABLE')
    }
}

export function concatenate(...arrays: (ArrayBuffer | Uint8Array)[]): ArrayBuffer {
    const buffersLengths = arrays.map((array) => array.byteLength)
    const totalLength = buffersLengths.reduce((p, c) => p + c, 0)
    const buffer = new Uint8Array(totalLength)
    buffersLengths.reduce(function (p, c, i) {
        const v = arrays[i]
        // @ts-ignore
        buffer.set(new Uint8Array(v.buffer ?? v), p)
        return p + c
    }, 0)
    return buffer.buffer
}

export function jwkToCOSEKey(jwk: JsonWebKey): ArrayBuffer {
    // todo: currently we only support ecdh p-256 algorithm
    const array: Uint8Array[] = []
    array.push(new Uint8Array([0xa5])) // size 5 of map
    array.push(new Uint8Array([0x01])) // key: kty
    array.push(new Uint8Array([0x02])) // value: EC
    array.push(new Uint8Array([0x20])) // key: crv
    array.push(new Uint8Array([0x01])) // value: P-256
    array.push(new Uint8Array([0x03])) // key: alg
    array.push(new Uint8Array([0x26])) // value: -7
    array.push(new Uint8Array(encode(-2))) // key: x
    array.push(new Uint8Array(encode(jwk.x)))
    array.push(new Uint8Array(encode(-3))) // key: y
    array.push(new Uint8Array(encode(jwk.y)))
    return concatenate(...array)
}

export function encodeAuthData(authData: AuthData): ArrayBuffer {
    // set idHash, 32 byte
    if (authData.rpIdHash.byteLength !== 32) {
        throw new TypeError('length of rpIdHash must be 32.')
    }
    // set flags, 1 byte
    const flagsBuffer = new Uint8Array(1)
    flagsBuffer.set([authData.flags | AuthDataFlag.AT | AuthDataFlag.UV], 0)
    // set signCount, 4 byte
    const signCountBuffer = new Uint32Array(1)
    let view = new DataView(signCountBuffer.buffer)
    view.setUint32(0, authData.signCount, false)
    // set attestedCredentialData
    const { credentialId, credentialPublicKey } = authData.attestedCredentialData
    const aaguidBuffer = new Uint32Array(4).fill(0) // is zero
    const credentialIdLengthBuffer = new Uint16Array(1)
    view = new DataView(credentialIdLengthBuffer.buffer)
    view.setUint16(0, credentialId.byteLength, false)
    const publicKeyBuffer = jwkToCOSEKey(credentialPublicKey)
    return concatenate(
        authData.rpIdHash,
        flagsBuffer.buffer,
        signCountBuffer.buffer,
        aaguidBuffer.buffer,
        credentialIdLengthBuffer.buffer,
        credentialId,
        publicKeyBuffer,
    )
}
