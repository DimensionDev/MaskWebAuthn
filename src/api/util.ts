import { decode } from 'cbor-redux'
import type { AttestationObject, PublicKeyCredential } from '../../types/interface'

export function concatenate(...arrays: (ArrayBuffer | Uint8Array)[]): ArrayBuffer {
    const buffersLengths = arrays.map((array) => array.byteLength)
    const totalLength = buffersLengths.reduce((p, c) => p + c, 0)
    const buffer = new Uint8Array(totalLength)
    buffersLengths.reduce(function (p, c, i) {
        const v = arrays[i]
        if (v instanceof ArrayBuffer) {
            buffer.set(new Uint8Array(v), p)
        } else {
            buffer.set(v, p)
        }
        return p + c
    }, 0)
    return buffer.buffer
}

export async function verify(publicKey: CryptoKey, credential: PublicKeyCredential): Promise<boolean> {
    const { clientDataJSON, attestationObject } = credential.response
    if (attestationObject) {
        const decodedAttestationObject = decode(attestationObject) as AttestationObject
        const {
            attStmt: { sig },
            authData,
        } = decodedAttestationObject
        const clientDataJsonHash = await crypto.subtle.digest('SHA-256', clientDataJSON)
        return crypto.subtle.verify(
            {
                name: 'ECDSA',
                hash: 'SHA-256',
            },
            publicKey,
            sig,
            concatenate(authData, clientDataJsonHash),
        )
    }
    // fallback
    return Promise.resolve().then(() => true)
}
