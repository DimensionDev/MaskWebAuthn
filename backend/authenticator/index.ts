import { concatenate, encodeAuthData, serializeCollectedClientData, sha256 } from '../util'
import { Buffer } from 'buffer'
import { encode } from 'cbor-redux'
import type { CollectedClientData, AttestationObject } from '../../types/interface'
import btoa from 'btoa'
import type { PublicKeyCredential } from '../../types/interface'
import { Alg } from '../../types/interface'

function getSignatureParams(alg: Alg): EcdsaParams {
    if (alg === Alg.ES256) {
        return {
            name: 'ECDSA',
            hash: 'SHA-256',
        }
    } else {
        throw new TypeError('Unsupported algorithm')
    }
}

const supportSet = new Set([Alg.ES256]) as ReadonlySet<number>

export async function generateCreationResponse(
    // backend creator provided
    credentialId: ArrayBuffer,
    keys: CryptoKeyPair,
    signCount: number,
    // RP provided
    rpID: string,
    clientData: CollectedClientData,
    algs: number[],
    signal?: AbortSignal,
): Promise<PublicKeyCredential> {
    const { publicKey, privateKey } = keys

    if (signal?.aborted) {
        throw new DOMException('AbortError')
    }

    const publicKeyJwk = await crypto.subtle.exportKey('jwk', publicKey)
    const rpHostname: ArrayBuffer = Buffer.from(new URL(rpID).hostname, 'utf-8').buffer
    const rpHash = await sha256(rpHostname)

    const authData = encodeAuthData({
        rpIdHash: rpHash,
        flags: 0,
        signCount,
        attestedCredentialData: {
            aaugid: '0', // we not support aaguid
            credentialId,
            credentialPublicKey: publicKeyJwk,
        },
        extensions: undefined,
    })

    const clientDataJson = serializeCollectedClientData({ ...clientData })
    const clientDataJsonBuffer = Buffer.from(clientDataJson)
    const clientDataJsonHash = await sha256(clientDataJsonBuffer)

    // start sign
    const signType = algs.find((alg) => supportSet.has(alg))
    if (!signType) {
        throw new Error('Unsupported algorithm')
    }
    const signParams = getSignatureParams(signType)
    const signTarget = concatenate(authData, clientDataJsonHash)
    const signature = await crypto.subtle.sign(signParams, privateKey, signTarget)
    // end sign

    const obj = {
        fmt: 'packed',
        attStmt: {
            alg: signType,
            sig: signature,
        },
        authData,
    }
    const attestationObject = encode<AttestationObject>(obj)

    return {
        type: 'public-key',
        id: btoa(String.fromCharCode(...new Uint8Array(credentialId))),
        rawId: credentialId,
        response: {
            clientDataJSON: clientDataJsonBuffer,
            attestationObject,
        },
    }
}
