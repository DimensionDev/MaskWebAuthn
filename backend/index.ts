import type { PublicKeyAuthenticatorProtocol } from '../types/interface'
import { create } from './publicKey'

export interface CreateAuthenticatorOptions {
  privateKey: JsonWebKey
  publicKey: JsonWebKey
}

export function createPublicKeyAuthenticator (opts: CreateAuthenticatorOptions): PublicKeyAuthenticatorProtocol {
  const { privateKey, publicKey } = opts
  const abortController = new AbortController()
  const abortSignal = abortController.signal

  return {
    // todo
    create: create.bind(undefined, privateKey, publicKey),
    signal: abortSignal,
    abort: () => abortController.abort(),
  }
}
