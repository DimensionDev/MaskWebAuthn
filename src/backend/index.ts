import { create, get } from './publicKey'
import type { PublicKeyAuthenticatorProtocol } from '../types/interface'

export interface CreateAuthenticatorOptions {
    // sign count
    getSignCount(key: CryptoKey, rpID: string, credentialID: ArrayBuffer): Promise<number>
    incrementSignCount(key: CryptoKey, rpID: string, credentialID: ArrayBuffer): Promise<void>

    hasCredential(rpID: string, credentialID: ArrayBuffer): Promise<boolean>

    // without username
    getResidentKeyPair(rpID: string): Promise<readonly [key: CryptoKeyPair, credentialID: ArrayBuffer]>

    // with username
    getKeyPairByKeyWrap(
        rpID: string,
        credentialIDs: BufferSource[],
    ): Promise<readonly [key: CryptoKeyPair | null, credentialID: ArrayBuffer]>
}

export function createPublicKeyAuthenticator(opts: CreateAuthenticatorOptions): PublicKeyAuthenticatorProtocol {
    return {
        create: create.bind(undefined, opts),
        get: get.bind(undefined, opts),
    }
}
