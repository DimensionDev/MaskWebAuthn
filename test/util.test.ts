import { concatenate, isRegistrableDomain } from '../backend/util'

function parseAuthData (buffer: ArrayBuffer) {
  const textDecoder = new TextDecoder()
  let rpIdHash = buffer.slice(0, 32)
  buffer = buffer.slice(32)
  textDecoder.decode(buffer)

  let flagsBuf = buffer.slice(0, 1)
  buffer = buffer.slice(1)
  let flagsInt = new Uint8Array(flagsBuf)[0]
  let flags = {
    up: !!(flagsInt & 0x01),
    uv: !!(flagsInt & 0x04),
    at: !!(flagsInt & 0x40),
    ed: !!(flagsInt & 0x80),
    flagsInt,
  }
  let counterBuf = buffer.slice(0, 4)
  buffer = buffer.slice(4)

  let bufferView = new DataView(counterBuf)
  let counter = bufferView.getUint32(0)
  let aaguid = undefined
  let credID = undefined
  let COSEPublicKey = undefined
  if (flags.at) {
    aaguid = buffer.slice(0, 16)
    buffer = buffer.slice(16)
    let credIDLenBuf = buffer.slice(0, 2)
    buffer = buffer.slice(2)
    bufferView = new DataView(credIDLenBuf)
    let credIDLen = bufferView.getUint16(0)
    credID = buffer.slice(0, credIDLen)
    buffer = buffer.slice(credIDLen)
    COSEPublicKey = buffer
  }
  return {
    rpIdHash,
    flagsBuf,
    flags,
    counter,
    counterBuf,
    aaguid,
    credID,
    COSEPublicKey,
  }
}

test('is registrable domain', () => {
  expect(isRegistrableDomain('', '')).toBe(false)
  expect(isRegistrableDomain('google.com', 'google.cn')).toBe(false)
  expect(isRegistrableDomain('http://google.com', 'https://google.com')).
    toBe(false)
  expect(isRegistrableDomain('https://foo.google.com', 'google.cn')).toBe(false)
  expect(isRegistrableDomain('https://google.com', 'google.com')).toBe(true)
  expect(isRegistrableDomain('https://google.com', 'https://google.cn')).
    toBe(false)
  expect(isRegistrableDomain('https://test.google.com', 'https://google.com')).
    toBe(true)
  expect(isRegistrableDomain('www.google.com', 'google.com')).toBe(true)
})

test('buffer contact check', () => {
  const a = new Uint8Array(4)
  const b = new Uint32Array(1)
  const c = new Uint16Array(2)
  a.set([0xf7, 0x7e, 0x7e, 0x7f], 0)
  b.set([0x3ff77ff3], 0)
  c.set([0x3ff3, 0x1122], 0)
  const res = concatenate(a.buffer, b.buffer, c.buffer)
  expect(res.byteLength).toBe(12)
  const dataView = new DataView(res)
  expect(dataView.getUint8(0)).toBe(0xf7)
  expect(dataView.getUint8(1)).toBe(0x7e)
  expect(dataView.getUint8(2)).toBe(0x7e)
  expect(dataView.getUint8(3)).toBe(0x7f)
  // 0x3f f7 7f f3
  expect(dataView.getUint8(4)).toBe(0xf3)
  expect(dataView.getUint8(5)).toBe(0x7f)
  expect(dataView.getUint8(6)).toBe(0xf7)
  expect(dataView.getUint8(7)).toBe(0x3f)
  // 0x3f f3
  expect(dataView.getUint8(8)).toBe(0xf3)
  expect(dataView.getUint8(9)).toBe(0x3f)
  expect(dataView.getUint8(10)).toBe(0x22)
  expect(dataView.getUint8(11)).toBe(0x11)
})
