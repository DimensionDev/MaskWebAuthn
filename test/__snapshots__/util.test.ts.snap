// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`serialize collected clientData: serialized collected client data 1`] = `"{\\"type\\":\\"webauthn.create\\",\\"challenge\\":{\\"type\\":\\"Buffer\\",\\"data\\":[0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15]},\\"origin\\":\\"https://google.com\\",\\"crossOrigin\\":false}"`;
